### ISSUE TITLE
Confirm and document whether or not `zonal_statistics` works for other geometries

### ISSUE THREAD

**Issue by emlys** [natcap/pygeoprocessing#322]
The docstring and implementation of `zonal_statistics` suggest it's meant for polygons, but it might work on other geometries too. I accidentally ran it on a point veector and got results, but I don't know if they're correct. We should confirm whether or not `zonal_statistics` produces valid results on other geometries. If so, update the documentation and tests to support this. If not, assert that the input geometries are indeed polygons.

**Comment by emlys** [comment]
There seem to be 3 components of the `zonal_statistics` logic that could be affected by different geometry types:
### 1. Calculating disjoint feature sets
Per the name of this function, `calculate_disjoint_polygon_set`, it was clearly designed for polygons. But it appears that it would work the same on any geometry type. The Rtree only cares about bounding boxes, which can be created for any geometry ([point bounding boxes are allowed](https://rtree.readthedocs.io/en/latest/tutorial.html#insert-records-into-the-index)). The shapely operations should work on any geometry.

This approach isn't the best for point geometries, though - if a point corresponds to a single pixel value, it could be more efficient to rasterize all unique points at once. Duplicate points would have the same stats, and multipoint stats could be calculated easily from the stats of their constituent points.

### 2. Rasterizing the features by FID
`gdal.Rasterize` works on any type of geometry. Note that the `ALL_TOUCHED` option (which we have discussed supporting https://github.com/natcap/pygeoprocessing/issues/312) is documented to work on lines and polygons; unclear what happens with points. I would hope it behaves like `ALL_TOUCHED=TRUE`, since it's unlikely that a point would happen to lie on the pixel centerpoint.
### 3. The edge case handling where we calculate stats for a feature's bounding box if the feature itself doesn't overlap any pixel centerpoints
This would technically still work, but it's much less likely to make sense with linestrings. Similarly to points, maybe we should use `ALL_TOUCHED=TRUE` for linestrings and avoid this edge case handling.

In conclusion, there's significant overlap between this issue and https://github.com/natcap/pygeoprocessing/issues/312. I'll need to try out rasterizing the different geometry types with and without `ALL_TOUCHED=TRUE` to understand what happens. But at a minimum, I think it's safe to say that `zonal_statistics` produces correct results on multipolygons, and we should update here: https://github.com/natcap/pygeoprocessing/blob/b50fead0314bbc15f2fb4b7c7c9c29667be6700f/src/pygeoprocessing/geoprocessing.py#L1712-L1715 to reflect that, and avoid the "inserting multipolygon into a polygon layer" warnings that we see in invest a lot.

**Comment by emlys** [comment]
To wrap up this ticket, I think `zonal_statistics`'s current algorithm and approach to rasterization doesn't make sense with linestring or point geometries. It _would_ make sense if we supported `ALL_TOUCHED=TRUE` (#312), but that would require a fundamentally different approach, because of the potential for more than 1 geometry to rasterize to the same pixel. 

Multipolygons should be supported, and we can update `zonal_statistics` to get rid of the warning on multipolygons. I think we should also check the geometry type and raise an error if it's not polygon or multipolygon.
